File: android/src/main/java/io/github/wangyng/simple_audio_player/SimpleAudioPlayerPlugin.java
package io.github.wangyng.simple_audio_player;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.github.wangyng.simple_audio_player.player.AudioFocusManager;
import io.github.wangyng.simple_audio_player.player.AudioNotificationManager;
import io.github.wangyng.simple_audio_player.player.ExoPlayerManager;
import io.github.wangyng.simple_audio_player.player.PlayerManager;
import io.github.wangyng.simple_audio_player.player.Song;


public class SimpleAudioPlayerPlugin implements FlutterPlugin, SimpleAudioPlayerApi, AudioFocusManager.AudioFocusChangeCallback {

    Map<Integer, PlayerManager> playerManagerMap = new HashMap<>();
    SimpleAudioPlayerEventSink songStateStream;

    AudioFocusManager audioFocusManager;
    SimpleAudioPlayerEventSink audioFocusStream;

    AudioNotificationManager audioNotificationManager;
    SimpleAudioPlayerEventSink notificationStream;

    SimpleAudioPlayerEventSink becomingNoisyStream;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        SimpleAudioPlayerApi.setup(binding, this, binding.getApplicationContext());
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        SimpleAudioPlayerApi.setup(binding, null, null);
    }

    @Override
    public void setSongStateStream(Context context, SimpleAudioPlayerEventSink songStateStream) {
        this.songStateStream = songStateStream;
    }

    @Override
    public void setAudioFocusStream(Context context, SimpleAudioPlayerEventSink audioFocusStream) {
        this.audioFocusStream = audioFocusStream;
    }

    @Override
    public void setNotificationStream(Context context, SimpleAudioPlayerEventSink notificationStream) {
        this.notificationStream = notificationStream;
    }

    @Override
    public void setBecomingNoisyStream(Context context, SimpleAudioPlayerEventSink becomingNoisyStream) {
        this.becomingNoisyStream = becomingNoisyStream;
    }

    @Override
    public void init(Context context, int playerId) {
        PlayerManager player = new ExoPlayerManager(context);
        player.setCallback(new PlayerManager.SongStateCallback() {

            @Override
            public void onReady() {
                if (songStateStream != null && songStateStream.event != null) {
                    Map<String, Object> result = new HashMap<>();
                    result.put("playerId", playerId);
                    result.put("event", "onReady");
                    result.put("data", player.getDuration());
                    songStateStream.event.success(result);
                }
            }

            @Override
            public void onPlayEnd() {
                if (songStateStream != null && songStateStream.event != null) {
                    Map<String, Object> result = new HashMap<>();
                    result.put("playerId", playerId);
                    result.put("event", "onPlayEnd");
                    songStateStream.event.success(result);
                }
            }

            @Override
            public void onError(@NotNull String message) {
                if (songStateStream != null && songStateStream.event != null) {
                    Map<String, Object> result = new HashMap<>();
                    result.put("playerId", playerId);
                    result.put("event", "onError");
                    result.put("data", message);
                    songStateStream.event.success(result);
                }
            }

            @Override
            public void onPositionChange(int position, int duration) {
                if (songStateStream != null && songStateStream.event != null) {
                    Map<String, Object> result = new HashMap<>();
                    result.put("playerId", playerId);
                    result.put("event", "onPositionChange");
                    result.put("data", new long[]{position, duration});
                    songStateStream.event.success(result);
                }
            }

            @Override
            public void onReceivePlay() {
                if (notificationStream.event != null) {
                    notificationStream.event.success("onPlay");
                }
            }

            @Override
            public void onReceivePause() {
                if (notificationStream.event != null) {
                    notificationStream.event.success("onPause");
                }
            }

            @Override
            public void onReceiveSkipToNext() {
                if (notificationStream.event != null) {
                    notificationStream.event.success("onSkipToNext");
                }
            }

            @Override
            public void onReceiveSkipToPrevious() {
                if (notificationStream.event != null) {
                    notificationStream.event.success("onSkipToPrevious");
                }
            }

            @Override
            public void onReceiveStop() {
                if (notificationStream.event != null) {
                    notificationStream.event.success("onStop");
                }
            }
        });
        playerManagerMap.put(playerId, player);

    }

    @Override
    public void prepare(Context context, int playerId, String uri) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            player.prepare(Uri.parse(uri));
        }
    }

    @Override
    public void play(Context context, int playerId) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            player.play();
        }
    }

    @Override
    public void pause(Context context, int playerId) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            player.pause();
        }
    }

    @Override
    public void stop(Context context, int playerId) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            player.stop();
        }
    }

    @Override
    public void seekTo(Context context, int playerId, int position) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            player.seekTo(position);
        }
    }

    @Override
    public void setVolume(Context context, int playerId, double volume) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            player.setVolume(volume);
        }
    }

    @Override
    public void setPlaybackRate(Context context, int playerId, double playbackRate) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            player.setRate(playbackRate);
        }
    }

    @Override
    public int getCurrentPosition(Context context, int playerId) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            return player.getCurrentPosition();
        } else {
            return 0;
        }
    }

    @Override
    public int getDuration(Context context, int playerId) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            return player.getDuration();
        } else {
            return 0;
        }
    }

    @Override
    public double getPlaybackRate(Context context, int playerId) {
        PlayerManager player = playerManagerMap.get(playerId);
        if (player != null) {
            return player.getPlaybackRate();
        } else {
            return 1;
        }
    }

    @Override
    public boolean tryToGetAudioFocus(Context context) {
        if (audioFocusManager == null) {
            audioFocusManager = new AudioFocusManager(context);
        }
        return audioFocusManager.tryToGetAudioFocus(this);
    }

    @Override
    public void giveUpAudioFocus(Context context) {
        if (audioFocusManager != null) {
            audioFocusManager.giveUpAudioFocus();
        }
    }

    @Override
    public void onAudioFocused() {
        if (audioFocusStream.event != null) {
            audioFocusStream.event.success("audioFocused");
        }
    }

    @Override
    public void onAudioNoFocus() {
        if (audioFocusStream.event != null) {
            audioFocusStream.event.success("audioNoFocus");
        }
    }

    @Override
    public void onAudioBecomingNoisy() {
        if (becomingNoisyStream.event != null) {
            becomingNoisyStream.event.success("becomingNoisy");
        }
    }

    @Override
    public void showNotification(Context context,int playerId, String title, String artist, String clipArt) {
        PlayerManager player = playerManagerMap.get(playerId);

        if (audioNotificationManager == null) {
            audioNotificationManager = new AudioNotificationManager(context);
            audioNotificationManager.setCallback(new AudioNotificationManager.AudioNotificationEventCallback() {
                @Override
                public void onReceivePlay() {
                    if (notificationStream.event != null) {
                        notificationStream.event.success("onPlay");
                    }
                }

                @Override
                public void onReceivePause() {
                    if (notificationStream.event != null) {
                        notificationStream.event.success("onPause");
                    }
                }

                @Override
                public void onReceiveSkipToNext() {
                    if (notificationStream.event != null) {
                        notificationStream.event.success("onSkipToNext");
                    }
                }

                @Override
                public void onReceiveSkipToPrevious() {
                    if (notificationStream.event != null) {
                        notificationStream.event.success("onSkipToPrevious");
                    }
                }

                @Override
                public void onReceiveStop() {
                    if (notificationStream.event != null) {
                        notificationStream.event.success("onStop");
                    }
                }
            });

            audioNotificationManager.showNotification(player, new Song(title, artist, clipArt));
        } else {
            audioNotificationManager.updateNotification(player, new Song(title, artist, clipArt));
        }
    }

    @Override
    public void cancelNotification(Context context) {
        if (audioNotificationManager != null) {
            audioNotificationManager.cancelNotification();
            audioNotificationManager = null;
        }
    }
}

File: ios/Classes/SimpleAudioPlayerPlugin.m
//
//  SimpleAudioPlayerPlugin.m
//  Pods
//
//  Created by 汪洋 on 2021/8/6.
//

#import "SimpleAudioPlayerPlugin.h"
#import "SimpleAudioPlayerEventSink.h"
#import "player/SimpleAudioPlayerManager.h"
#import "player/SimpleAudioFocusManager.h"
#import "player/SimpleAudioNotificationManager.h"

@interface SimpleAudioPlayerPlugin () <SimpleAudioFocusChangeDelegate, SimpleAudioNotificationDelegate>

@property(nonatomic, strong) NSMutableDictionary *playerManagerMap;
@property(nonatomic, strong) NSMutableDictionary *playerManagerDelegateMap;
@property(nonatomic, strong) SimpleAudioPlayerEventSink *songStateStream;

@property(nonatomic, strong) SimpleAudioFocusManager *audioFocusManager;
@property(nonatomic, strong) SimpleAudioPlayerEventSink *audioFocusStream;

@property(nonatomic, strong) SimpleAudioNotificationManager *notificationManager;
@property(nonatomic, strong) SimpleAudioPlayerEventSink *notificationStream;

@property(nonatomic, strong) SimpleAudioPlayerEventSink *becomingNoisyStream;

@end

@interface SimpleAudioPlayerStateImpl : NSObject<SimpleAudioPlayerStateDelegate>

@property(nonatomic, weak) SimpleAudioPlayerPlugin* plugin;

@property(nonatomic, assign) NSInteger playerId;

@end

@implementation SimpleAudioPlayerPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    SimpleAudioPlayerPlugin* instance = [[SimpleAudioPlayerPlugin alloc] init];
    [SimpleAudioPlayerApi setup:registrar api:instance];
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _playerManagerMap = [[NSMutableDictionary alloc] init];
        _playerManagerDelegateMap = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)initWithPlayerId:(NSInteger)playerId {
    SimpleAudioPlayerManager *player = [[SimpleAudioPlayerManager alloc] init];
    SimpleAudioPlayerStateImpl *playerDelegate = [[SimpleAudioPlayerStateImpl alloc] init];
    playerDelegate.plugin = self;
    playerDelegate.playerId = playerId;
    player.delegate = playerDelegate;
    self.playerManagerMap[@(playerId)] = player;
    self.playerManagerDelegateMap[@(playerId)] = playerDelegate;
}

- (void)prepareWithPlayerId:(NSInteger)playerId uri:(NSString *)uri {
    SimpleAudioPlayerManager *player = self.playerManagerMap[@(playerId)];
    NSURL *url;
    if ([uri hasPrefix:@"http"]) {
        url = [[NSURL alloc] initWithString:uri];
    } else if ([uri hasPrefix:@"file://"]){
        url = [[NSURL alloc] initFileURLWithPath:[uri substringFromIndex:@"file://".length]];
    }
    [player prepareWithUrl:url];
}

- (void)playWithPlayerId:(NSInteger)playerId {
    SimpleAudioPlayerManager *player = self.playerManagerMap[@(playerId)];
    [player play];
}

- (void)pauseWithPlayerId:(NSInteger)playerId {
    SimpleAudioPlayerManager *player = self.playerManagerMap[@(playerId)];
    [player pause];
}

- (void)stopWithPlayerId:(NSInteger)playerId {
    SimpleAudioPlayerManager *player = self.playerManagerMap[@(playerId)];
    [player stop];
}

- (void)seekToWithPlayerId:(NSInteger)playerId position:(NSInteger)position {
    SimpleAudioPlayerManager *player = self.playerManagerMap[@(playerId)];
    [player seekToWithPosition:position];
}

- (void)setVolumeWithPlayerId:(NSInteger)playerId volume:(double)volume {
    SimpleAudioPlayerManager *player = self.playerManagerMap[@(playerId)];
    [player setVolume:volume];
}

- (NSInteger)getCurrentPositionWithPlayerId:(NSInteger)playerId {
    SimpleAudioPlayerManager *player = self.playerManagerMap[@(playerId)];
    return player.currentPosition;
}

- (NSInteger)getDurationWithPlayerId:(NSInteger)playerId {
    SimpleAudioPlayerManager *player = self.playerManagerMap[@(playerId)];
    return player.duration;
}

- (void)giveUpAudioFocus {
    [self.audioFocusManager giveUpAudioFocus];
}

- (BOOL)tryToGetAudioFocus {
    if (self.audioFocusManager == nil) {
        self.audioFocusManager = [[SimpleAudioFocusManager alloc] init];
    }
    self.audioFocusManager.delegate = self;
    return [self.audioFocusManager tryToGetAudioFocus];
}

- (void)onAudioFocused {
    if (self.audioFocusStream.event) {
        self.audioFocusStream.event(@"audioFocused");
    }
}

- (void)onAudioNoFocus {
    if (self.audioFocusStream.event) {
        self.audioFocusStream.event(@"audioNoFocus");
    }
}

- (void)onAudioBecomingNoisy {
    if (self.becomingNoisyStream.event) {
        self.becomingNoisyStream.event(@"becomingNoisy");
    }
}

- (void)showNotificationWithTitle:(NSString *)title artist:(NSString *)artist clipArt:(NSString *)clipArt {
    if (self.notificationManager == nil) {
        self.notificationManager = [[SimpleAudioNotificationManager alloc] init];
        self.notificationManager.delegate = self;
    }
    SimpleAudioPlayerSong *song = [[SimpleAudioPlayerSong alloc] initWithTitle:title artist:artist clipArt:clipArt];
    [self.notificationManager showNotificationWithSong:song];
}

- (void)updateNotificationWithShowPlay:(BOOL)showPlay title:(NSString *)title artist:(NSString *)artist clipArt:(NSString *)clipArt {
    SimpleAudioPlayerSong *song = [[SimpleAudioPlayerSong alloc] initWithTitle:title artist:artist clipArt:clipArt];
    [self.notificationManager updateNotificationWithShowPlay:showPlay song:song];
}

- (void)cancelNotification {
    [self.notificationManager cancelNotification];
}

- (void)onReceivePause {
    if (self.notificationStream.event) {
        self.notificationStream.event(@"onPause");
    }
}

- (void)onReceivePlay {
    if (self.notificationStream.event) {
        self.notificationStream.event(@"onPlay");
    }
}

- (void)onReceiveSkipToNext {
    if (self.notificationStream.event) {
        self.notificationStream.event(@"onSkipToNext");
    }
}

- (void)onReceiveSkipToPrevious {
    if (self.notificationStream.event) {
        self.notificationStream.event(@"onSkipToPrevious");
    }
}

- (void)onReceiveStop {
    if (self.notificationStream.event) {
        self.notificationStream.event(@"onStop");
    }
}

@end

@implementation SimpleAudioPlayerStateImpl

- (void)onReadyWithDuration:(NSInteger)duration; {
    if (self.plugin.songStateStream.event) {
        NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
        result[@"playerId"] = @(self.playerId);
        result[@"event"] = @"onReady";
        result[@"data"] = @(duration);
        self.plugin.songStateStream.event(result);
    }
}

- (void)onPlayEnd {
    if (self.plugin.songStateStream.event) {
        NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
        result[@"playerId"] = @(self.playerId);
        result[@"event"] = @"onPlayEnd";
        self.plugin.songStateStream.event(result);
    }
}


- (void)onErrorWithMessage:(NSString *)message {
    if (self.plugin.songStateStream.event) {
        NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
        result[@"playerId"] = @(self.playerId);
        result[@"event"] = @"onError";
        result[@"data"] = message;
        self.plugin.songStateStream.event(result);
    }
}

- (void)onPositionChangeWithPosition:(NSInteger)position duration:(NSInteger)duration {
    if (self.plugin.songStateStream.event) {
        NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
        result[@"playerId"] = @(self.playerId);
        result[@"event"] = @"onPositionChange";
        result[@"data"] = @[@(position), @(duration)];
        self.plugin.songStateStream.event(result);
    }
}


@end
